### ISSUE TITLE
kube-controller-manager crash: invalid memory address  or nil pointer dereference

### ISSUE THREAD

**Issue by krotz-dieter** [kubernetes/kubernetes#128935]
### What happened?

Enabling ingress nginx in version v1.31.2 causes this error stack:
```
I1122 11:02:33.848730       1 job_controller.go:568] "enqueueing job" logger="job-controller" key="ingress-nginx/ingress-nginx-admission-patch" delay="1s"
E1122 11:02:34.022503       1 panic.go:261] "Observed a panic" panic="runtime error: invalid memory address or nil pointer dereference" panicGoValue="\"invalid memory address or nil pointer dereference\"" stacktrace=<
        goroutine 912 [running]:
        k8s.io/apimachinery/pkg/util/runtime.logPanic({0x38483f0, 0x554eb20}, {0x2d9c260, 0x54897b0})
                k8s.io/apimachinery/pkg/util/runtime/runtime.go:107 +0xbc
        k8s.io/apimachinery/pkg/util/runtime.handleCrash({0x38483f0, 0x554eb20}, {0x2d9c260, 0x54897b0}, {0x554eb20, 0x0, 0x43d945?})
                k8s.io/apimachinery/pkg/util/runtime/runtime.go:82 +0x5e
        k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc001a31180?})
                k8s.io/apimachinery/pkg/util/runtime/runtime.go:59 +0x108
        panic({0x2d9c260?, 0x54897b0?})
                runtime/panic.go:770 +0x132
        k8s.io/cloud-provider/controllers/service.(*Controller).needsUpdate(0xc000c491e0, 0xc0022ffb88, 0xc00240cc88)
                k8s.io/cloud-provider/controllers/service/controller.go:586 +0x39a
        k8s.io/cloud-provider/controllers/service.New.func2({0x32711e0?, 0xc0022ffb88?}, {0x32711e0, 0xc00240cc88?})
                k8s.io/cloud-provider/controllers/service/controller.go:144 +0x74
        k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
                k8s.io/client-go/tools/cache/controller.go:253
        k8s.io/client-go/tools/cache.(*processorListener).run.func1()
                k8s.io/client-go/tools/cache/shared_informer.go:976 +0xea
        k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
                k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x33
        k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00258ef70, {0x38135c0, 0xc001a08c90}, 0x1, 0xc001a0aa80)
                k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xaf
        k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc001a39770, 0x3b9aca00, 0x0, 0x1, 0xc001a0aa80)
                k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x7f
        k8s.io/apimachinery/pkg/util/wait.Until(...)
                k8s.io/apimachinery/pkg/util/wait/backoff.go:161
        k8s.io/client-go/tools/cache.(*processorListener).run(0xc000d327e0)
                k8s.io/client-go/tools/cache/shared_informer.go:972 +0x69
        k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
                k8s.io/apimachinery/pkg/util/wait/wait.go:72 +0x52
        created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start in goroutine 553
                k8s.io/apimachinery/pkg/util/wait/wait.go:70 +0x73
 >
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x211fb1a]

goroutine 912 [running]:
k8s.io/apimachinery/pkg/util/runtime.handleCrash({0x38483f0, 0x554eb20}, {0x2d9c260, 0x54897b0}, {0x554eb20, 0x0, 0x43d945?})
        k8s.io/apimachinery/pkg/util/runtime/runtime.go:89 +0xee
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc001a31180?})
        k8s.io/apimachinery/pkg/util/runtime/runtime.go:59 +0x108
panic({0x2d9c260?, 0x54897b0?})
        runtime/panic.go:770 +0x132
k8s.io/cloud-provider/controllers/service.(*Controller).needsUpdate(0xc000c491e0, 0xc0022ffb88, 0xc00240cc88)
        k8s.io/cloud-provider/controllers/service/controller.go:586 +0x39a
k8s.io/cloud-provider/controllers/service.New.func2({0x32711e0?, 0xc0022ffb88?}, {0x32711e0, 0xc00240cc88?})
        k8s.io/cloud-provider/controllers/service/controller.go:144 +0x74
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
        k8s.io/client-go/tools/cache/controller.go:253
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
        k8s.io/client-go/tools/cache/shared_informer.go:976 +0xea
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
        k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x33
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc001e63f70, {0x38135c0, 0xc001a08c90}, 0x1, 0xc001a0aa80)
        k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xaf
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc001a39770, 0x3b9aca00, 0x0, 0x1, 0xc001a0aa80)
        k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x7f
k8s.io/apimachinery/pkg/util/wait.Until(...)
        k8s.io/apimachinery/pkg/util/wait/backoff.go:161
k8s.io/client-go/tools/cache.(*processorListener).run(0xc000d327e0)
        k8s.io/client-go/tools/cache/shared_informer.go:972 +0x69
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
        k8s.io/apimachinery/pkg/util/wait/wait.go:72 +0x52
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start in goroutine 553
        k8s.io/apimachinery/pkg/util/wait/wait.go:70 +0x73
```

### What did you expect to happen?

kube-controller-manager does not crash and restart

### How can we reproduce it (as minimally and precisely as possible)?

- download latest [K2s](https://github.com/Siemens-Healthineers/K2s)
- on windows 10/11 call 
`k2s install` 
(check  [prerequisites](https://github.com/Siemens-Healthineers/K2s/blob/main/docs/op-manual/installing-k2s.md#prerequisites)) in order to create the kubernetes cluster
- the enable the ingress nginx with: 
`k2s addons enable ingress nginx`
- after last step look at crashed kube controller manager with
`k logs kube-controller-manager-kubemaster -n kube-system -p`

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```
Client Version: v1.31.2
Kustomize Version: v5.4.2
Server Version: v1.31.2
</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
$ uname -a
Linux kubemaster 5.10.0-33-cloud-amd64 #1 SMP Debian 5.10.226-1 (2024-10-03) x86_64 GNU/Linux

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
```
BuildNumber  Caption                          OSArchitecture  Version
26100        Microsoft Windows 11 Enterprise  64-bit          10.0.26100
</details>

### Install tools

<details>
[K2s](https://github.com/Siemens-Healthineers/K2s)
</details>

### Container runtime (CRI) and version (if applicable)

<details>
not applicable
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
not applicable
</details>

**Comment by krotz-dieter** [comment]
We are just using internally kubeadm. It's actually easy to set up on a Windows machine.

**Comment by liggitt** [comment]
I think the event recorder is getting initialized in Run:

https://github.com/kubernetes/kubernetes/blob/5864a4677267e6adeae276ad85882a8714d69d9d/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L224

but update handlers can be called earlier as soon as informers are started

@pohly @aojea who may have seen similar init order issues

**Comment by liggitt** [comment]
Looks like this was relocated in https://github.com/kubernetes/kubernetes/commit/50c12437604b0cd5a73514389409fc2fde8b91bd and introduced the panic possibility in 1.31+

/kind regression
/assign @pohly 
/priority important-soon

**Comment by pohly** [comment]
This got fixed in https://github.com/kubernetes/kubernetes/pull/128179. @aojea advised [against backporting that](https://github.com/kubernetes/kubernetes/pull/128179#issuecomment-2430463198) because the fix is part of a larger cleanup of the cloud service controller.

Shall we backport or revert the part of https://github.com/kubernetes/kubernetes/pull/122145 that caused this?

**Comment by liggitt** [comment]
> revert the part of https://github.com/kubernetes/kubernetes/pull/122145 that caused this

I'd be more inclined to narrowly revert the bit that could cause panics in these controllers for 1.31

**Comment by pohly** [comment]
I agree => https://github.com/kubernetes/kubernetes/pull/128959

**Comment by aojea** [comment]
@pohly @liggitt this is fixed by https://github.com/kubernetes/kubernetes/pull/128236/commits/fb63fe6926e3341976a63ccc9ce49c870eaa2f5e in 1.31.2

@krotz-dieter are we sure this is with 1.31.2?
is the kube-controller-manager setting a cloud provider ? otherwise the service lb controller will never be initialized

**Comment by pohly** [comment]
> download latest [K2s](https://github.com/Siemens-Healthineers/K2s)

Please give the exact version. "Latest" tends to change over time...

Currently it's [v1.1.3](https://github.com/Siemens-Healthineers/K2s/releases/tag/v1.1.3), released well before Kubernetes [v1.31.3 which contains the backport](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.31.md#changelog-since-v1312).

Looks like you need a new release of K2s.

**Comment by pohly** [comment]
> this is fixed by https://github.com/kubernetes/kubernetes/commit/fb63fe6926e3341976a63ccc9ce49c870eaa2f5e in 1.31.2

That commit got released only in 1.31.3, not 1.31.2. But thanks for pointing it out, that helps decide whether further work is needed - probably not.

**Comment by aojea** [comment]
I see, you are right, I got confused by looking at the git history I thought it made it for 1.31.2
```
commit 5864a4677267e6adeae276ad85882a8714d69d9d (tag: v1.31.2)
Author: Kubernetes Release Robot <k8s-release-robot@users.noreply.github.com>
Date:   Tue Oct 22 20:28:12 2024

    Release commit for Kubernetes v1.31.2

commit fb63fe6926e3341976a63ccc9ce49c870eaa2f5e
Author: carlory <baofa.fan@daocloud.io>
Date:   Fri Oct 18 10:27:12 2024

    Fix crash on kube manager's service-lb-controller after v1.31.0.

    Co-authored-by: Antonio Ojea <antonio.ojea.garcia@gmail.com>

```

**Comment by krotz-dieter** [comment]
Will test it with 1.31.3, thanks for the fast responses !

**Comment by krotz-dieter** [comment]
Tested with 1.31.3, issue is gone. Many thanks !

**Comment by pohly** [comment]
Thanks for the confirmation @krotz-dieter.

/close

**Comment by uhthomas** [comment]
I know this has been fixed in 1.31.3 - I just wanted to say I also observed this today in 1.31.1 on bare metal (Talos). The stack looks the same. Just really random.

```sh
E0213 20:21:42.098176       1 panic.go:261] "Observed a panic" panic="runtime error: invalid memory address or nil pointer dereference" panicGoValue="\"invalid memory address or nil pointer dereference\"" stacktrace=<
        goroutine 2003 [running]:
        k8s.io/apimachinery/pkg/util/runtime.logPanic({0x38433c0, 0x5545b00}, {0x2d99020, 0x5480790})
                k8s.io/apimachinery/pkg/util/runtime/runtime.go:107 +0xbc
        k8s.io/apimachinery/pkg/util/runtime.handleCrash({0x38433c0, 0x5545b00}, {0x2d99020, 0x5480790}, {0x5545b00, 0x0, 0x43d945?})
                k8s.io/apimachinery/pkg/util/runtime/runtime.go:82 +0x5e
        k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc002621500?})
                k8s.io/apimachinery/pkg/util/runtime/runtime.go:59 +0x108
        panic({0x2d99020?, 0x5480790?})
                runtime/panic.go:770 +0x132
        k8s.io/cloud-provider/controllers/service.(*Controller).needsUpdate(0xc000a32340, 0xc0053a9688, 0xc006030288)
                k8s.io/cloud-provider/controllers/service/controller.go:562 +0x728
        k8s.io/cloud-provider/controllers/service.New.func2({0x326c980?, 0xc0053a9688?}, {0x326c980, 0xc006030288?})
                k8s.io/cloud-provider/controllers/service/controller.go:144 +0x74
        k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
                k8s.io/client-go/tools/cache/controller.go:253
        k8s.io/client-go/tools/cache.(*processorListener).run.func1()
                k8s.io/client-go/tools/cache/shared_informer.go:976 +0xea
        k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
                k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x33
        k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc003787f70, {0x380e820, 0xc002115410}, 0x1, 0xc000ae6c60)
                k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xaf
        k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00262c770, 0x3b9aca00, 0x0, 0x1, 0xc000ae6c60)
                k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x7f
        k8s.io/apimachinery/pkg/util/wait.Until(...)
                k8s.io/apimachinery/pkg/util/wait/backoff.go:161
        k8s.io/client-go/tools/cache.(*processorListener).run(0xc000fe77a0)
                k8s.io/client-go/tools/cache/shared_informer.go:972 +0x69
        k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
                k8s.io/apimachinery/pkg/util/wait/wait.go:72 +0x52
        created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start in goroutine 1796
                k8s.io/apimachinery/pkg/util/wait/wait.go:70 +0x73
 >
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x211c928]

goroutine 2003 [running]:
k8s.io/apimachinery/pkg/util/runtime.handleCrash({0x38433c0, 0x5545b00}, {0x2d99020, 0x5480790}, {0x5545b00, 0x0, 0x43d945?})
        k8s.io/apimachinery/pkg/util/runtime/runtime.go:89 +0xee
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc002621500?})
        k8s.io/apimachinery/pkg/util/runtime/runtime.go:59 +0x108
panic({0x2d99020?, 0x5480790?})
        runtime/panic.go:770 +0x132
k8s.io/cloud-provider/controllers/service.(*Controller).needsUpdate(0xc000a32340, 0xc0053a9688, 0xc006030288)
        k8s.io/cloud-provider/controllers/service/controller.go:562 +0x728
k8s.io/cloud-provider/controllers/service.New.func2({0x326c980?, 0xc0053a9688?}, {0x326c980, 0xc006030288?})
        k8s.io/cloud-provider/controllers/service/controller.go:144 +0x74
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
        k8s.io/client-go/tools/cache/controller.go:253
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
        k8s.io/client-go/tools/cache/shared_informer.go:976 +0xea
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
        k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x33
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc007b97f70, {0x380e820, 0xc002115410}, 0x1, 0xc000ae6c60)
        k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xaf
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00262c770, 0x3b9aca00, 0x0, 0x1, 0xc000ae6c60)
        k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x7f
k8s.io/apimachinery/pkg/util/wait.Until(...)
        k8s.io/apimachinery/pkg/util/wait/backoff.go:161
k8s.io/client-go/tools/cache.(*processorListener).run(0xc000fe77a0)
        k8s.io/client-go/tools/cache/shared_informer.go:972 +0x69
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
        k8s.io/apimachinery/pkg/util/wait/wait.go:72 +0x52
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start in goroutine 1796
        k8s.io/apimachinery/pkg/util/wait/wait.go:70 +0x73
```

**Comment by aojea** [comment]
yeah, sorry about that 🙇
